modules/ding_react/ding_react.module

```diff
@@ -0,0 +1,371 @@
+    ['query' => ['destination' => current_path()]]
+  );
+}
+
+function ding_react_admin_settings_form() {
+  $form = [];
+
+  $form['services'] = [
+    '#type' => 'fieldset',
+    '#title' => t('Services')
+  ];
+
+  $form['services']['ding_react_material_list_url'] = [
+    '#type' => 'textfield',
+    '#title' => t('Material List'),
+    '#description' => t('Url to the Material List service instance to use.'),
+    '#default_value' => variable_get('ding_react_material_list_url', ''),
+    '#required' => TRUE,
+    '#element_validate' => [ 'ding_react_element_validate_url' ],
+  ];
+
+  $form['services']['ding_react_follow_searches_url'] = [
+    '#type' => 'textfield',
+    '#title' => t('Follow Searches'),
+    '#description' => t('Url to the Follow Searches service instance to use.'),
+    '#default_value' => variable_get('ding_react_follow_searches_url', ''),
+    '#required' => TRUE,
+    '#element_validate' => [ 'ding_react_element_validate_url' ],
+  ];
+
+  $form['services']['ding_react_migrate_timeout'] = [
+    '#type' => 'textfield',
+    '#title' => t('Migration timeout'),
+    '#description' => t('The number of seconds to wait for an external service to complete a migration during login. A high timeout reduces risk of the migration not completing. A low timeout prevents the login procedure from stalling. Incomplete migrations will be retried the next time the user logs in.'),
+    '#default_value' => variable_get('ding_react_migrate_timeout', 5),
+    '#required' => TRUE,
+    '#element_validate' => [ 'element_validate_integer_positive' ],
+  ];
+
+  return system_settings_form($form);
+}
+
+function ding_react_element_validate_url($element, &$form_state, $form) {
+  if (!empty($element['#value']) && !valid_url($element['#value'], true)) {
+    form_error($element, t('Please enter a valid url.'));
+  }
+}
+
+/**
+ * Migrate a user account from using a legacy user id to new token-based one.
+ *
+ * @param \stdClass $account
+ *   The user account to migrate.
+ */
+function ding_react_user_migrate(stdClass $account) {
+  $migrated_prefix = 'migrated-';
+
+  $legacy_uid = $account->openlist_uid;
+
+  try {
+    $token = ding_provider_invoke('openplatform_token', 'get');
+  } catch (DingProviderNoProvider $e) {
+    $token = NULL;
+  } catch (DingProviderDoesntImplement $e) {
+    $token = NULL;
+  }
+
+  if (!empty($legacy_uid)
+    && !preg_match("/${migrated_prefix}/", $legacy_uid)
+    && !empty($token)) {
+    $migrate_request_options = [
+      'method' => 'PUT',
+      // Use a relatively low timeout. We do not what login to stall if services
+      // are slow to respond. Then migration will have to wait.
+      'timeout' => variable_get('ding_react_migrate_timeout', 5),
+      'headers' => [
+        'Authorization' => "Bearer ${token}",
+      ]
+    ];
+
+    $material_list_url = variable_get('ding_react_material_list_url');
+    $response = drupal_http_request("${material_list_url}/migrate/${legacy_uid}",
```

Comment:
dunno if foo("${bar}/migrate/${bas}") is better than foo($bar . '/migrate/' . $bas)? Made me pause, anyway...